ddl/column.go

```diff
@@ -532,53 +644,144 @@ func checkDropColumnForStatePublic(tblInfo *model.TableInfo, colInfo *model.Colu
 	return nil
 }
 
-func onDropColumn(t *meta.Meta, job *model.Job) (ver int64, _ error) {
-	tblInfo, colInfo, idxInfos, err := checkDropColumn(t, job)
+func (w *worker) onDropColumn(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
+	tblInfo, colInfo, sidxInfos, cidxInfos, err := checkDropColumn(t, job)
 	if err != nil {
 		return ver, errors.Trace(err)
 	}
 
-	originalState := colInfo.State
+	if len(cidxInfos) > 0 && job.IsRollingback() {
+		// Handle the rolling back job.
+		ctidxInfos := getTempCompositeIndexes(tblInfo, cidxInfos)
+		ver, err = doDropIndexes(t, job, tblInfo, ctidxInfos)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		return ver, nil
+	}
+
+	colOriginalState := colInfo.State
+	// Handle reorg composite indices
+	if len(cidxInfos) > 0 && colOriginalState == model.StatePublic {
```

Comment:
For drop column with an index on it, if the index has multiple columns, the process is like this:
```
idx1 (a, b, c)  drop column b 
-> drop index idx1 -> drop column b > create index idx1(a, c)
```
Will we record the whole state changes in the schema diff history or we just record the `drop column` schema diff change?
The pessimistic transaction commit amend relies on the schema diff change history, it will try to generate mutations based on the schema change diff contents. If there is only a `drop column` schema diff change in the schema diff history queue, the current implementation of the transaction amend will treat it as amendable, and it will not try to do the `create index idx1(a, c)` related mutations generation so that the amend result will be incorrect.
@wjhuang2016 may have some advice on this compatibility problem.

Reply:
if we use drop then create index, the rollingback process will be hard to handle.